Remove support for this.attrs (#1016)
emberjs/rfcs: text/1016-deprecate-this-attrs-in-templates.md

---

> stage: recommended

Reviewer: Can't start in recommended?

Author: I copied this from the original RFC, apologies

Author: done

start-date: 2020-12-22T00:00:00.000Z
release-date: 2021-03-22T00:00:00.000Z
release-versions:
  ember-source: v3.26.0

teams:
  - framework
prs:
  accepted: https://github.com/emberjs/rfcs/pull/1016 
project-link:
---

# Deprecate using `{{attrs}}` and {{this.attrs}} in templates

## Summary

[RFC #690](https://github.com/emberjs/rfcs/pull/690) had some unfortunate timing where `{{attrs}}` was widely used, but before usage of `{{this.x}}` became widely common. As such, the deprecation messaging [implemented here](https://github.com/emberjs/ember.js/blob/0822efd9c0554f11d01232585596ac70c14a4306/packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts#L58) only messaged in the case of `{{attrs}}`, rather than both `{{attrs}}` and `{{this.attrs}}`. 

The RFC had intended, and even describes in its transition path to move away from `{{this.attrs}}`. [The code for the build time assertion](https://github.com/emberjs/ember.js/blob/0822efd9c0554f11d01232585596ac70c14a4306/packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts#L84) even tried to handle the `{{this.attrs}}` case. 

```hbs
{{attrs.foo}} and {{this.attrs.foo}}

> {{! is equivalent to }}
> {{@foo}} and {{@foo}}

Reviewer 2: I would find this less confusing

```suggestion
{{! are equivalent to }}
{{@foo}}
```

```

It was a legacy API that existed prior to named arguments being introduced in
Ember, and has continued to be supported via a template transform for some time.

> This RFC proposes that we immediately remove support for `{{this.attrs}}` in favor of named arguments.

Reviewer: So this is proposing we remove support without issuing a deprecation in a minor release?

I _think_ that is what you are proposing cause otherwise we could implement the deprecation from the earlier RFC.

Author: Since 6.0 is so close, I can change this phrasing to deprecate for removal in v6

Author: or rather, to fix the build-time assertion to warn until v6

Author: updated

> This RFC proposes that we add a deprecation warning to the build-time assertion that forbids `{{attrs}}`, it will also forbid `{{this.attrs}}`. 

Reviewer 2: ```suggestion
This RFC proposes that we add a deprecation warning to the build-time assertion that forbids `{{attrs}}` and `{{this.attrs}}`. 
```
